kategory-docs/docs/docs/datatypes/statet/README.md

```diff
@@ -6,4 +6,142 @@ permalink: /docs/datatypes/statet/
+```
+
+Luckily Kategory offers some nice solutions [`Functional Error Handling` docs]({{ '/docs/patterns/error_handling' | relative_url }}).
+Now we can model our error domain with ease.
+
+```kotlin:ank:silent
+sealed class StackError {
+    object StackEmpty : StackError()
+}
+
+typealias StackEmpty = StackError.StackEmpty
+
+fun popE(stack: Stack): Either<StackError, Tuple2<Stack, String>> =
+        if (stack.isEmpty()) StackEmpty.left()
+        else stack.first().let {
+            stack.drop(1) toT it
+        }.right()
+
+fun pushE(s: String, stack: Stack): Either<StackError, Tuple2<Stack, Unit>> =
+        (listOf(s, *stack.toTypedArray()) toT Unit).right()
+
+fun stackOperationsE(stack: Stack): Either<StackError, Tuple2<Stack, String>> {
+    return pushE("a", stack).flatMap { (s1, _) ->
+        popE(s1).flatMap { (s2, _) ->
+            popE(s2)
+        }
+    }
+}
+```
+```kotlin:ank
+stackOperationsE(listOf("hello", "world", "!"))
+```
+```kotlin:ank
+stackOperationsE(listOf())
+```
+
+As is immediately clear, this code while properly modelling the errors, has become more complex but our signature now represents a simple `Stack` as a `List` with an error domain.
+Let's refactor our manual state management in the form of `(S) -> Tuple2<S, A>` to `State`.
+
+So what we want is a return type that represents `Either` a `StackError` or a certain `State` of `Stack.` When working with `State` we don't pass around `Stack` anymore, so there is no parameter to check if the `Stack` is empty.
+
+```kotlin:ank:silent
+fun _popS(): Either<StackError, StateT<IdHK, Stack, String>> = TODO()
+```
+
+The only thing we can do is handle this with `StateT`. We want to wrap `State` with `Either`. Left to right wrap `State` `T` (with) `Either` `StateT<EitherHK`.
+`EitherKindPartial` is an alias that helps us to fix `StackError` as the left type parameter for `Either<L, R>`.
+
+```kotlin:ank
+fun popS() = StateT<EitherKindPartial<StackError>, Stack, String> { stack ->
+    if (stack.isEmpty()) StackEmpty.left()
+    else stack.first().let {
+        stack.drop(1) toT it
+    }.right()
+}
+
+fun pushS(s: String) = StateT<EitherKindPartial<StackError>, Stack, Unit> { stack ->
+    (listOf(s, *stack.toTypedArray()) toT Unit).right()
+}
+
+fun stackOperationsS(): StateT<EitherKindPartial<StackError>, Stack, String> {
+    return pushS("a").flatMap({ _ ->
+        popS().flatMap({ _ ->
+            popS()
+        }, Either.monad())
+    }, Either.monad()).ev()
+}
+
+stackOperationsS().runM(listOf("hello", "world", "!"))
+```
+```kotlin:ank
+stackOperationsS().runM(listOf())
+```
+
+While our code looks very similar to what we had before there are some key advantages. State management is now contained within `State` and we are dealing only with 1 monad instead of 2 nested monads so we can use monad bindings!
+
+```kotlin:ank
+fun stackOperationsS2() = StateT .monad<EitherKindPartial<StackError>, Stack>().binding {
```

Comment:
pretty helpful to know that transformers enable monad comprenhensions after collapsing everything to a single type 👏